Lesson19/elifbasak/script.js

```diff
@@ -0,0 +1,161 @@
+   array `#items` to store the cart items.
+3. Add a `viewCart` method to display all items in the cart.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 2: Add Items to the Cart
+-----------------------------------------------------------
+1. Create an `addItem` method in the `ShoppingCart` class.
+2. The method should:
+   - Accept `name`, `price`, and `quantity` as parameters.
+   - Check if the item already exists in the cart.
+     - If it exists, increase the quantity.
+     - Otherwise, add the new item to the `#items` array.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 3: Remove Items from the Cart
+-----------------------------------------------------------
+1. Add a `removeItem` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept the `name` of the item to remove.
+   - Remove the item from the `#items` array if it exists.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 4: Calculate the Total Cost
+-----------------------------------------------------------
+1. Add a `getTotal` method to the `ShoppingCart` class.
+2. The method should:
+   - Calculate and return the total cost of all items in 
+     the cart.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 5: Apply a Discount
+-----------------------------------------------------------
+1. Add an `applyDiscount` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept a discount code (e.g., 'SAVE10', 'SAVE20').
+   - Apply a percentage discount to the total cost if the 
+     code is valid.
+3. Use an object to store discount codes and their values.
+*/
+
+class ShoppingCart {
+  #items;
+
+  constructor() {
+    this.#items = [];
+  }
+
+  viewCart() {
+    console.log("Viewing cart....");
+    for (const item of this.#items) {
+      console.log(item);
+    }
+  }
+
+  addItem(name, price, quantity) {
+    console.log("Adding new item to the cart...");
+    for (const item of this.#items) {
+      if (name === item.name) {
+        console.log("increase the quantity.");
+        item.quantity++;
+        return;
+      }
+    }
+
+    const id = new Date().getMilliseconds();
+    this.#items.push({ id, name, price, quantity });
+  }
+
+  // In real life, we usually use id to find elements or do any manipulations
+  removeItem(name) {
+    for (let i = 0; i < this.#items.length; i++) {
+      const currentElement = this.#items[i];
+      if (name.toLowerCase() === currentElement.name.toLowerCase()) {
+        this.#items.splice(i, 1);
+        return;
+      }
+    }
+  }
+
+  getTotal() {
+    console.log("Calculating the total cost of the cart....");
+    let total = 0;
+    for (const item of this.#items) {
+      total += item.quantity * item.price.amount;
+    }
+    console.log(`Total amount of the cart is ${total} TRY.`);
+    return total;
+  }
+
+  applyDiscount({ discountCode, discountPercentage }) {
```

Comment:
Define discount codes as an object and then check that the discountCode provided when calling this function is in the object. If it's there, then apply it to the total of the cart, otherwise print a message saying that discount code is not valid. 🙂